File: src/echo_activity.erl
-module(echo_activity).

% elp:ignore W0012 W0040 E1599
-moduledoc {file, "../docs/echo_activity.md"}.

-export([
    execute/2,
    terminate/1,
    handle_heartbeat/1,
    handle_cancel/1,
    handle_message/2
]).

-include_lib("temporal_sdk/include/activity.hrl").

-doc false.
execute(#{task := #{attempt := Attempt}}, [_Echo, SleepTime, FailedAttempts]) when
    is_integer(SleepTime), is_integer(FailedAttempts), Attempt < FailedAttempts
->
    timer:sleep(SleepTime),
    throw("Test echo_activity throw");
execute(_Context, [Echo, SleepTime, FailedAttempts]) when
    is_integer(SleepTime), is_integer(FailedAttempts)
->
    timer:sleep(SleepTime),
    [do_echo(Echo)];
execute(_Context, [Echo, SleepTime]) when is_integer(SleepTime) ->
    timer:sleep(SleepTime),
    [do_echo(Echo)];
execute(_Context, [Echo]) ->
    [do_echo(Echo)];
execute(_Context, Echo) ->
    do_echo(Echo).

do_echo(Echo) when is_binary(Echo), byte_size(Echo) > 10_000 -> ~"large_payload";
do_echo(Echo) when is_list(Echo), length(Echo) > 3_300 -> ~"large_payload";
do_echo(Echo) -> Echo.

-doc false.
handle_heartbeat(_Context) ->
    heartbeat.

-doc false.
handle_cancel(#{cancel_requested := true}) ->
    {cancel, ["cancel details"]};
handle_cancel(_Context) ->
    ignore.

-doc false.
handle_message(_Context, _Message) ->
    ignore.

-doc false.
terminate(_Context) ->
    ok.

File: lib/echo_activity.ex
defmodule EchoActivity do
  use TemporalSdk.Activity
  @moduledoc false

  @impl true
  def execute(%{task: %{attempt: attempt}}, [_echo, sleep_time, failed_attempts])
      when is_integer(sleep_time) and is_integer(failed_attempts) and attempt < failed_attempts do
    Process.sleep(sleep_time)
    throw("Test echo_activity throw")
  end

  def execute(_context, [echo, sleep_time, failed_attempts])
      when is_integer(sleep_time) and is_integer(failed_attempts) do
    Process.sleep(sleep_time)
    [do_echo(echo)]
  end

  def execute(_context, [echo, sleep_time]) when is_integer(sleep_time) do
    Process.sleep(sleep_time)
    [do_echo(echo)]
  end

  def execute(_context, [echo]), do: [do_echo(echo)]
  def execute(_context, echo), do: do_echo(echo)

  @impl true
  def handle_heartbeat(_context), do: :heartbeat

  @impl true
  def handle_cancel(%{cancel_requested: true}), do: {:cancel, ["cancel details"]}
  def handle_cancel(_context), do: :ignore

  @impl true
  def handle_message(_context, _message), do: :ignore

  @impl true
  def terminate(_context), do: :ok

  defp do_echo(echo) when is_binary(echo) and byte_size(echo) > 10_000, do: <<"large_payload">>
  defp do_echo(echo) when is_list(echo) and length(echo) > 3_300, do: <<"large_payload">>
  defp do_echo(echo), do: echo
end
